File: Python/learn/ReAct_agent.py
# 定义工具
from tools.baidu_search import BaiduSearchTool
from tools.file_manage import current_time
tools = [BaiduSearchTool(), current_time]
tools_by_name = {tool.name: tool for tool in tools}
# print(tools_by_name)

# 定义大模型
from langchain.chat_models import init_chat_model
model = init_chat_model(
    model="deepseek-chat",
    model_provider="deepseek",
    temperature=0.7,
    max_tokens=None,
    timeout=None,
    # base_url="https://api.deepseek.com",
    max_retries=2,
)

# 绑定工具
model_with_tools = model.bind_tools(tools)


# 定义状态
from langchain.messages import AnyMessage
from typing_extensions import Annotated, TypedDict
import operator

# 在 LangGraph 中，状态会在整个智能体的执行期间持续存在。
# 使用 operator.add 声明的 Annotated 类型能确保新消息被追加到现有列表中，而不是将其替换。
class MessagesState(TypedDict):
    """消息状态"""
    messages: Annotated[list[AnyMessage], operator.add]
    llm_calls: int
    tools_calls: int


# 定义节点

# 大模型调用节点
from langchain.messages import SystemMessage


def llm_call(state: dict):
    """LLM decides whether to call a tool or not"""
    messages = state["messages"]
    response = model_with_tools.invoke(
        [
            SystemMessage(
                content="You are a helpful assistant tasked with performing arithmetic on a set of inputs."
            )
        ]
        + messages
    )
    return {
        "messages": [response],
        "llm_calls": state.get('llm_calls', 0) + 1
    }
# 工具调用节点
from langchain.messages import ToolMessage

def tool_node(state: dict):
    """Performs the tool call"""
    result = []
    for tool_call in state["messages"][-1].tool_calls:
        tool = tools_by_name[tool_call["name"]]
        observation = tool.invoke(tool_call["args"])
        result.append(ToolMessage(content=observation, tool_call_id=tool_call["id"]))
    return {"messages": result}


# 定义条件边
from typing import Literal
from langgraph.graph import StateGraph, START, END


def should_continue(state: MessagesState) -> Literal["tool_node", END]:
    """Decide if we should continue the loop or stop based upon whether the LLM made a tool call"""
    messages = state["messages"]
    last_message = messages[-1]

    if last_message.tool_calls:
        return "tool_node"

    return END

# 定义工作流
graph = StateGraph(MessagesState)
graph.add_node(llm_call)
graph.add_node(tool_node)
graph.add_edge(START, "llm_call")
graph.add_conditional_edges("llm_call", should_continue, ["tool_node", END])
graph.add_edge("tool_node", "llm_call")

# 编译工作流
agent = graph.compile()



# 调用工作流
from langchain.messages import HumanMessage
messages = [HumanMessage(content="今天深圳天气怎么样？出行如何穿搭？")]
messages = agent.invoke({"messages": messages})
for m in messages["messages"]:
    m.pretty_print()

File: TypeScript/learn/agents/ReAct_agent.ts
// 定义工具
import { BaiduSearchTool } from "../tools/baiduSearchTool";
import { getCurrentTimeTool } from "../tools/common";

// 定义大模型
import { ChatDeepSeek } from "@langchain/deepseek";

const model = new ChatDeepSeek({
	model: "deepseek-chat",
	temperature: 0.8,
});

//
const baiduTool = new BaiduSearchTool();
const toolsByName = {
	[baiduTool.name]: baiduTool,
	[getCurrentTimeTool.name]: getCurrentTimeTool,
};

// 绑定工具
const model_with_tools = model.bindTools([baiduTool, getCurrentTimeTool]);

// 定义状态
import * as z from "zod";
import { BaseMessage } from "@langchain/core/messages";
import { MessagesZodMeta } from "@langchain/langgraph";
import { registry } from "@langchain/langgraph/zod";

const MessagesState = z.object({
	messages: z
		.array(z.custom<BaseMessage>())
		.register(registry, MessagesZodMeta),
	llmCalls: z.number().optional(),
});

import { SystemMessage } from "@langchain/core/messages";

const systemPrompt = `
# 角色
你是一个专业的RAG（检索增强生成）智能助手，具备强大的信息检索和生成能力。

# 核心要求
1. **信息检索**：使用搜索baidu_search工具获取最新、权威的客观信息和知识
2. **信息时效性**：请时刻使用get_current_time工具获取当前时间，确保信息时效性
3. **信息准确性**：生成内容请以搜索结果为准，避免臆测和幻觉，确保信息准确性
`;
// 定义大模型节点
const llm_call_node = async (state: z.infer<typeof MessagesState>) => {
	console.log("llm_call_node");
	const result = await model_with_tools.invoke([
		new SystemMessage(systemPrompt),
		...state.messages,
	]);
	return {
		messages: [...state.messages, result],
		llmCalls: (state.llmCalls || 0) + 1,
	};
};

import { ToolMessage, AIMessage } from "@langchain/core/messages";
// 定义工具节点
const tool_call_node = async (state: z.infer<typeof MessagesState>) => {
	console.log("tool_call_node");
	const lastMessage = state.messages[state.messages.length - 1];
	if (!(lastMessage instanceof AIMessage)) {
		return {
			messages: state.messages,
		};
	}
	const result: ToolMessage[] = [];
	for (const toolCall of lastMessage.tool_calls ?? []) {
		const tool = toolsByName[toolCall.name];
		const observation = await tool.invoke(toolCall);
		result.push(observation);
	}
	return {
		messages: [...result],
	};
};

// 定义工具调用条件边
const shouldContinue = async (state: z.infer<typeof MessagesState>) => {
	const lastMessage = state.messages[state.messages.length - 1];
	if (!(lastMessage instanceof AIMessage)) {
		return "END";
	}
	if (lastMessage.tool_calls?.length) {
		return "toolCall";
	}
	return "END";
};

// 构建工作流
import { StateGraph, START, END } from "@langchain/langgraph";
import { MemorySaver } from "@langchain/langgraph";

const workflow = new StateGraph(MessagesState)
	.addNode("llmCall", llm_call_node)
	.addNode("toolCall", tool_call_node)
	.addEdge(START, "llmCall")
	.addConditionalEdges("llmCall", shouldContinue, {
		toolCall: "toolCall",
		END: END,
	})
	.addEdge("toolCall", "llmCall");

const checkpointer = new MemorySaver();

const agent = workflow.compile({
	checkpointer,
});

const config = {
	configurable: { thread_id: "1" },
};

import { HumanMessage } from "@langchain/core/messages";
// const invoke = async () => {
// 	const result = await agent.invoke({
// 		messages: [new HumanMessage({ content: "请帮我写一个关于机器学习的博客" })],
// 	});
// 	// console.log(result);
// 	if (result && result.messages && result.messages.length > 0) {
// 		console.log(result.messages[result.messages.length - 1].content);
// 	}
// };
// invoke();

// 交互式CLI：模仿 Python while True 输入循环
import * as readLine from "node:readline";
async function startCli(): Promise<void> {
	const rl = readLine.createInterface({
		input: process.stdin,
		output: process.stdout,
		terminal: true,
	});

	console.log("请输入搜索关键词，输入 quit/exit/bye 退出");
	rl.setPrompt("User: ");
	rl.prompt();

	rl.on("line", async (line: string) => {
		const userInput = line.trim();
		if (["quit", "exit", "bye"].includes(userInput.toLowerCase())) {
			console.log("Goodbye!");
			rl.close();
			return;
		}
		try {
			const result = await agent.invoke(
				{
					messages: [new HumanMessage({ content: userInput })],
				},
				config
			);
			// console.log(result);
			if (result && result.messages && result.messages.length > 0) {
				console.log(result.messages[result.messages.length - 1].content);
			}
		} catch (err: any) {
			console.log(`'${userInput}' 调用失败：${err?.message ?? String(err)}`);
		}
		rl.prompt();
	});

	rl.on("close", () => {
		// 与 Python 示例一致，退出时结束进程
		if (typeof process !== "undefined") {
			process.exit(0);
		}
	});
}
startCli();
